compiler/qsc_eval/src/lib.rs

```diff
@@ -607,346 +620,157 @@ impl State {
+        // In that case, the values are empty and the result is unit.
+        self.val_register.take().unwrap_or_else(Value::unit)
     }
 
     #[allow(clippy::similar_names)]
-    fn cont_expr(
+    fn eval_expr(
         &mut self,
         env: &mut Env,
+        sim: &mut impl Backend<ResultType = impl Into<val::Result>>,
         globals: &impl PackageStoreLookup,
+        out: &mut impl Receiver,
         expr: ExprId,
     ) -> Result<(), Error> {
         let expr = globals.get_expr((self.package, expr).into());
         self.current_span = expr.span;
 
         match &expr.kind {
-            ExprKind::Array(arr) => self.cont_arr(arr),
-            ExprKind::ArrayRepeat(item, size) => self.cont_arr_repeat(globals, *item, *size),
-            ExprKind::Assign(lhs, rhs) => self.cont_assign(*lhs, *rhs),
-            ExprKind::AssignOp(op, lhs, rhs) => self.cont_assign_op(env, globals, *op, *lhs, *rhs),
-            ExprKind::AssignField(record, field, replace) => {
-                self.cont_assign_field(*record, field, *replace);
+            ExprKind::Array(arr) => self.eval_arr(arr.len()),
+            ExprKind::ArrayLit(arr) => self.eval_arr_lit(arr, globals),
+            ExprKind::ArrayRepeat(..) => self.eval_arr_repeat(expr.span)?,
+            ExprKind::Assign(lhs, _) => self.eval_assign(env, globals, *lhs)?,
+            ExprKind::AssignOp(op, lhs, rhs) => {
+                let rhs_span = globals.get_expr((self.package, *rhs).into()).span;
+                let (is_array, is_unique) =
+                    is_updatable_in_place(env, globals.get_expr((self.package, *lhs).into()));
+                if is_array {
+                    if is_unique {
+                        self.eval_array_append_in_place(env, globals, *lhs)?;
+                        return Ok(());
+                    }
+                    let rhs_val = self.take_val_register();
+                    self.eval_expr(env, sim, globals, out, *lhs)?;
+                    self.push_val();
+                    self.set_val_register(rhs_val);
+                }
+                self.eval_binop(*op, rhs_span)?;
+                self.eval_assign(env, globals, *lhs)?;
             }
-            ExprKind::AssignIndex(lhs, mid, rhs) => {
-                self.cont_assign_index(env, globals, *lhs, *mid, *rhs);
+            ExprKind::AssignField(record, field, _) => {
+                self.eval_update_field(field.clone());
+                self.eval_assign(env, globals, *record)?;
             }
-            ExprKind::BinOp(op, lhs, rhs) => self.cont_binop(globals, *op, *rhs, *lhs),
-            ExprKind::Block(block) => self.push_block(env, globals, *block),
+            ExprKind::AssignIndex(lhs, mid, _) => {
+                let mid_span = globals.get_expr((self.package, *mid).into()).span;
+                let (_, is_unique) =
+                    is_updatable_in_place(env, globals.get_expr((self.package, *lhs).into()));
```

Comment:
Along the same lines than my previous comment, would it be worth not discarding the `is_array` value and adding something like `assert!(is_array)` to make the assumptions more evident?

Reply:
This one I don't think is necessary, because `AssignIndex` is an expression that can only ever make sense for arrays.

Reply:
It's worth noting that `assert!` is not free. Unlike other languages where asserts are only compiled out by default, in Rust this check will still run in release unless you use the `debug_assert!` macro instead. We've avoid debug asserts so far, though I can't remember the motivation for why, perhaps just that we don't get much protection from them since we don't regularly run in debug mode even in tests.